Lib/concurrent/futures/process.py

```diff
@@ -360,7 +360,7 @@ def run(self):
                 if executor := self.executor_reference():
                     if process_exited:
                         with self.shutdown_lock:
-                            executor._adjust_process_count()
+                            executor._adjust_process_count(force=True)
                     else:
                         executor._idle_worker_semaphore.release()
```

Comment:
I think we should think more generally about what `_idle_worker_semaphore` is supposed to mean.

If the process has not exited, we release `_idle_worker_semaphore`, implying that the process was marked busy and now switches to idle.
If the process has exited, we call `_adjust_process_count` which *acquires* `_idle_worker_semaphore` as if the process had switched from idle to busy... but it has not (it was already considered busy).

So it seems to me that, if the process has exited, we should simply *not* acquire the semaphore as the process was not accounted as idle.

This would change the code to something like:
```python
    def _require_process(self):
        """
        Require a process for task execution, potentially spawning a new one.
        """
        # if there's an idle process, we don't need to spawn a new one.
        if not self._idle_worker_semaphore.acquire(blocking=False):
            self._adjust_process_count()

    def _adjust_process_count(self):
        ...
```

Reply:
And the call `_require_process` from `submit`, and call `_adjust_process_count` when a worker process exited.

Reply:
`_idle_worker_semaphore` is **meaningless** as long as we reach the maximum number of workers because the math would be **wrong** since then. Each `submit` tries to aqcuire the semaphore, but with `blocking=False`, which means `submit` will not necessarily decrement the counter. However, all the finished tasks will release the semaphore so all of them will increase the counter and the counter will just be something huge - the semaphore is useless at that point any any check against that is meaningless. `acquire` will always succeed and no new process will be spawn.

The math behind `_idle_worker_semaphore` is simply fragile and it only works when we have not reached max worker number.

Reply:
Ok... so the fix should address the problem of `_idle_worker_semaphore` being meaningless, IMHO.

`_idle_worker_semaphore` is not really used for synchronization since we never block on it. We merely use it as some kind of atomic integer. So perhaps we should replace that semaphore with a lock-protected integer (for lack of an atomic integer primitive in the Python stdlib). That would allow its value to become negative, and the accounting would become correct.

Reply:
Yes, we can do that. `threading.Semaphore` has a private `_value` attribute to access the value - which is what I originally mentioned that could make this work, but I'm hesitated to use it even internally. We can convert the whole `_idle_worker_semaphore` and make it always valid. Do you want me to do that instead?

Reply:
Let's not rely on the private `_value` attribute indeed.

Reply:
Okay I modified the implementation to make `_idle_worker_number` always represent what it means literally. I moved the "decrement" part out of `_adjust_process_count()` because that should not be part of the logic - and also when the process exits it should not decrease that. There are multiple ways to deal with the counter and I chose the one that I believe is easiest to follow.

Reply:
Do you think we should avoid increasing the counter when `_safe_to_dynamically_spawn_children == False`? We are just increasing it for no reason. Probably would save some time.

Reply:
> Do you think we should avoid increasing the counter when `_safe_to_dynamically_spawn_children == False`?

Well, as noted in the comments, we might still spawn children dynamically in that case, if some workers died for some reason. We can just keep the counter for simplicity.

Reply:
Unfortunately that won't work. Because the counter will only be decremented if `_safe_to_dynamically_spawn_children == True`, So `_adjust_process_count` will do nothing if a forked worker dead. We need plenty of extra effort to make it work. However, the original `force == True` way actually works in this scenario.

Reply:
> So _adjust_process_count will do nothing if a forked worker dead.

That sounds wrong and that also contradicts the longish comment in `_adjust_process_count` ("Assertion disabled as this codepath ...").

Reply:
Concretely, it seems that:
1. if a worker exits, `_adjust_process_count` is called and may launch another worker (in which case it will increment the count). That's regardless of the start method.
2. if a task is submitted, `_adjust_process_count` is called and the worker count is updated _only_ if `_safe_to_dynamically_spawn_children` is true.

It seems that creates a problem: if `_safe_to_dynamically_spawn_children` is false, the accounting of idle workers will break when workers exit and new tasks are submitted.

The solution should simply be to remove the `_safe_to_dynamically_spawn_children` condition in `submit`.

Reply:
> The solution should simply be to remove the _safe_to_dynamically_spawn_children condition in submit.

That would be wrong as it will spawn a process in `fork` case before `_start_executor_manager_thread` initialize the process pool. The original reason why this is added: https://github.com/python/cpython/issues/90622

Do we want to go deep on this matter? Because we need to decide what's the best option:
1. Risk dead-lock to always saturate the process pool
2. Risk the process pool to be unsaturated

We can't avoid 2. entirely because we relies on the return message from the worker process and the process could die before it can send the message.

Also removing the condition for good has performance impact on every submit in fork case, but may not matter that much.

Reply:
Also, besides the fact a new process will be spawned, the math is still incorrect because with fork, all the worker processes will be spawned without the counter calculation - we need to change that as well if we want to make it work. How large of an impact do we want this PR to be? Do we want to fix the whole function for good? If so, we probably need some extra design discussions, like should we check if the process pool is idle if a process dies in fork case?

BTW, again, my first patch actually happens to solve this issue :)

Reply:
> That sounds wrong and that also contradicts the longish comment in `_adjust_process_count` ("Assertion disabled as this codepath ...").

And yes, that comment has been wrong since this optimization was added. It was never correct due to the incorrect math.